// @tarojs/mini-runner/dist/quickapp/style/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDeclarationValue = exports.addComment = exports.addDeclaration = exports.removeDeclaration = exports.comment = void 0;
const comment = (msg, url) => 'Taro comment: ' + msg + '.' + (url ? '[兼容写法参考](' + url + ')' : '');
exports.comment = comment;
/**
 * 在rule中删除指定declaration
 */
const removeDeclaration = (declaration, rule) => rule.declarations.splice(rule.declarations.findIndex(v => v.property === declaration.property && v.value === declaration.value), 1);
exports.removeDeclaration = removeDeclaration;
/**
 * 在rule中增加一条declaration
 */
const addDeclaration = (property, value, rule) => rule.declarations.push({
    type: 'declaration',
    property,
    value
});
exports.addDeclaration = addDeclaration;
/**
 * 在rule中增加一条comment
 */
const addComment = (property, value, rule) => rule.declarations.push({
    type: 'comment',
    comment: `${property}:${value}; ${exports.comment('unsupported style', 'https://doc.quickapp.cn/widgets/common-styles.html')}`
});
exports.addComment = addComment;
const getDeclarationValue = (property, rule) => {
    const declarations = rule.declarations.filter(declaration => declaration.property === property);
    return (declarations.length && declarations[0]) || false;
};
exports.getDeclarationValue = getDeclarationValue;
