// @tarojs/mini-runner/dist/webpack/build.conf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const helper_1 = require("@tarojs/helper");
const chain_1 = require("./chain");
const base_conf_1 = require("./base.conf");
const MiniPlugin_1 = require("../plugins/MiniPlugin");
const vue_1 = require("./vue");
const vue3_1 = require("./vue3");
exports.default = (appPath, mode, config) => {
    const chain = base_conf_1.default(appPath);
    const { buildAdapter = helper_1.PLATFORMS.WEAPP, alias = {}, entry = {}, output = {}, fileType = {
        style: '.wxss',
        config: '.json',
        script: '.js',
        templ: '.wxml'
    }, globalObject = 'wx', outputRoot = 'dist', sourceRoot = 'src', isBuildPlugin = false, runtimePath, taroComponentsPath, designWidth = 750, deviceRatio, enableSourceMap = process.env.NODE_ENV !== 'production', sourceMapType, debugReact = false, baseLevel = 16, framework = 'nerv', prerender, minifyXML = {}, defineConstants = {}, env = {}, cssLoaderOption = {}, sassLoaderOption = {}, lessLoaderOption = {}, stylusLoaderOption = {}, mediaUrlLoaderOption = {}, fontUrlLoaderOption = {}, imageUrlLoaderOption = {}, miniCssExtractPluginOption = {}, postcss = {}, nodeModulesPath, isBuildQuickapp = false, template, quickappJSON, csso, terser, commonChunks, addChunkPages, blended, isBuildNativeComp, modifyMiniConfigs, modifyBuildAssets, onCompilerMake } = config;
    let { copy } = config;
    const plugin = {};
    const minimizer = [];
    const sourceDir = path.join(appPath, sourceRoot);
    const outputDir = path.join(appPath, outputRoot);
    const taroBaseReg = /@tarojs[\\/][a-z]+/;
    if (isBuildPlugin) {
        const patterns = copy ? copy.patterns : [];
        patterns.push({
            from: path.join(sourceRoot, 'plugin', 'doc'),
            to: path.join(outputRoot, 'doc')
        });
        patterns.push({
            from: path.join(sourceRoot, 'plugin', 'plugin.json'),
            to: path.join(outputRoot, 'plugin', 'plugin.json')
        });
        copy = Object.assign({}, copy, { patterns });
    }
    if (copy) {
        plugin.copyWebpackPlugin = chain_1.getCopyWebpackPlugin({ copy, appPath });
    }
    alias[helper_1.taroJsComponents + '$'] = taroComponentsPath || `${helper_1.taroJsComponents}/mini`;
    if (framework === 'react') {
        alias['react-dom$'] = '@tarojs/react';
        if (process.env.NODE_ENV !== 'production' && !debugReact) {
            alias['react-reconciler$'] = 'react-reconciler/cjs/react-reconciler.production.min.js';
            // eslint-disable-next-line dot-notation
            alias['react$'] = 'react/cjs/react.production.min.js';
            // eslint-disable-next-line dot-notation
            alias['scheduler$'] = 'scheduler/cjs/scheduler.production.min.js';
            alias['react/jsx-runtime$'] = 'react/cjs/react-jsx-runtime.production.min.js';
        }
    }
    if (framework === 'nerv') {
        alias['react-dom'] = 'nervjs';
        alias.react = 'nervjs';
    }
    env.FRAMEWORK = JSON.stringify(framework);
    env.TARO_ENV = JSON.stringify(buildAdapter);
    const constantsReplaceList = chain_1.mergeOption([chain_1.processEnvOption(env), defineConstants]);
    const entryRes = chain_1.getEntry({
        sourceDir,
        entry,
        isBuildPlugin
    });
    const defaultCommonChunks = isBuildPlugin
        ? ['plugin/runtime', 'plugin/vendors', 'plugin/taro', 'plugin/common']
        : ['runtime', 'vendors', 'taro', 'common'];
    let customCommonChunks = defaultCommonChunks;
    if (typeof commonChunks === 'function') {
        customCommonChunks = commonChunks(defaultCommonChunks.concat()) || defaultCommonChunks;
    }
    else if (Array.isArray(commonChunks) && commonChunks.length) {
        customCommonChunks = commonChunks;
    }
    plugin.definePlugin = chain_1.getDefinePlugin([constantsReplaceList]);
    const miniPluginOptions = {
        sourceDir,
        outputDir,
        constantsReplaceList,
        nodeModulesPath,
        isBuildQuickapp,
        template,
        fileType,
        quickappJSON,
        designWidth,
        pluginConfig: entryRes.pluginConfig,
        isBuildPlugin: Boolean(isBuildPlugin),
        commonChunks: customCommonChunks,
        baseLevel,
        framework,
        prerender,
        addChunkPages,
        modifyMiniConfigs,
        modifyBuildAssets,
        onCompilerMake,
        minifyXML,
        runtimePath,
        blended,
        isBuildNativeComp,
        alias
    };
    plugin.miniPlugin = !isBuildNativeComp ? chain_1.getMiniPlugin(miniPluginOptions) : chain_1.getBuildNativePlugin(miniPluginOptions);
    plugin.miniCssExtractPlugin = chain_1.getMiniCssExtractPlugin([{
            filename: `[name]${fileType.style}`,
            chunkFilename: `[name]${fileType.style}`
        }, miniCssExtractPluginOption]);
    plugin.providerPlugin = chain_1.getProviderPlugin({
        window: ['@tarojs/runtime', 'window'],
        document: ['@tarojs/runtime', 'document'],
        navigator: ['@tarojs/runtime', 'navigator'],
        requestAnimationFrame: ['@tarojs/runtime', 'requestAnimationFrame'],
        cancelAnimationFrame: ['@tarojs/runtime', 'cancelAnimationFrame'],
        Element: ['@tarojs/runtime', 'TaroElement']
    });
    const isCssoEnabled = !((csso && csso.enable === false));
    const isTerserEnabled = !((terser && terser.enable === false));
    if (mode === 'production') {
        if (isTerserEnabled) {
            minimizer.push(chain_1.getTerserPlugin([
                enableSourceMap,
                terser ? terser.config : {}
            ]));
        }
        if (isCssoEnabled) {
            const cssoConfig = csso ? csso.config : {};
            plugin.cssoWebpackPlugin = chain_1.getCssoWebpackPlugin([cssoConfig]);
        }
    }
    chain.merge({
        mode,
        devtool: chain_1.getDevtool(enableSourceMap, sourceMapType),
        entry: entryRes.entry,
        output: chain_1.getOutput(appPath, [{
                outputRoot,
                publicPath: '/',
                globalObject
            }, output]),
        target: MiniPlugin_1.createTarget({
            framework
        }),
        resolve: { alias },
        module: chain_1.getModule(appPath, {
            sourceDir,
            buildAdapter,
            isBuildQuickapp,
            // constantsReplaceList,
            designWidth,
            deviceRatio,
            enableSourceMap,
            compile: config.compile || {},
            cssLoaderOption,
            lessLoaderOption,
            sassLoaderOption,
            stylusLoaderOption,
            fontUrlLoaderOption,
            imageUrlLoaderOption,
            mediaUrlLoaderOption,
            postcss,
            fileType
        }),
        plugin,
        optimization: {
            usedExports: true,
            minimizer,
            runtimeChunk: {
                name: isBuildPlugin ? 'plugin/runtime' : 'runtime'
            },
            splitChunks: {
                chunks: 'all',
                maxInitialRequests: Infinity,
                minSize: 0,
                cacheGroups: {
                    common: {
                        name: isBuildPlugin ? 'plugin/common' : 'common',
                        minChunks: 2,
                        priority: 1
                    },
                    vendors: {
                        name: isBuildPlugin ? 'plugin/vendors' : 'vendors',
                        minChunks: 2,
                        test: module => {
                            return /[\\/]node_modules[\\/]/.test(module.resource);
                        },
                        priority: 10
                    },
                    taro: {
                        name: isBuildPlugin ? 'plugin/taro' : 'taro',
                        test: module => {
                            return taroBaseReg.test(module.context);
                        },
                        priority: 100
                    }
                }
            }
        }
    });
    switch (framework) {
        case helper_1.FRAMEWORK_MAP.VUE:
            vue_1.customVueChain(chain);
            break;
        case helper_1.FRAMEWORK_MAP.VUE3:
            vue3_1.customVue3Chain(chain);
            break;
        default:
    }
    return chain;
};
