// @tarojs/plugin-platform-swan/dist/components-react.js
const View = 'view';
const Icon = 'icon';
const Progress = 'progress';
const RichText = 'rich-text';
const Text = 'text';
const Button = 'button';
const Checkbox = 'checkbox';
const CheckboxGroup = 'checkbox-group';
const Form = 'form';
const Input = 'input';
const Label = 'label';
const Picker = 'picker';
const PickerView = 'picker-view';
const PickerViewColumn = 'picker-view-column';
const Radio = 'radio';
const RadioGroup = 'radio-group';
const Slider = 'slider';
const Switch = 'switch';
const CoverImage = 'cover-image';
const Textarea = 'textarea';
const CoverView = 'cover-view';
const MovableArea = 'movable-area';
const MovableView = 'movable-view';
const ScrollView = 'scroll-view';
const Swiper = 'swiper';
const SwiperItem = 'swiper-item';
const Navigator = 'navigator';
const Audio = 'audio';
const Camera = 'camera';
const Image = 'image';
const LivePlayer = 'live-player';
const Video = 'video';
const Canvas = 'canvas';
const Ad = 'ad';
const WebView = 'web-view';
const Block = 'block';
const Map = 'map';
const Slot = 'slot';
const CustomWrapper = 'custom-wrapper';

// For React.createElement's type
const Tabs = 'tabs';
const TabItem = 'tab-item';
const AnimationVideo = 'animation-video';
const AnimationView = 'animation-view';
const ArCamera = 'ar-camera';
const RtcRoom = 'rtc-room';
const RtcRoomItem = 'rtc-room-item';
const OpenData = 'open-data';

export { Ad, AnimationVideo, AnimationView, ArCamera, Audio, Block, Button, Camera, Canvas, Checkbox, CheckboxGroup, CoverImage, CoverView, CustomWrapper, Form, Icon, Image, Input, Label, LivePlayer, Map, MovableArea, MovableView, Navigator, OpenData, Picker, PickerView, PickerViewColumn, Progress, Radio, RadioGroup, RichText, RtcRoom, RtcRoomItem, ScrollView, Slider, Slot, Swiper, SwiperItem, Switch, TabItem, Tabs, Text, Textarea, Video, View, WebView };
