// @tarojs/components/dist/esm/taro-canvas-core.entry.js
import { r as registerInstance, c as createEvent, h, g as getElement } from './index-310de3d5.js';

const indexCss = "taro-canvas-core{position:relative;display:block;width:300px;height:150px}";

const LONG_TAP_DELAY = 500;
const Canvas = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.onLongTap = createEvent(this, "longtap", 7);
    this.onTouchStart = () => {
      this.timer = setTimeout(() => {
        this.onLongTap.emit();
      }, LONG_TAP_DELAY);
    };
    this.onTouchMove = () => {
      clearTimeout(this.timer);
    };
    this.onTouchEnd = () => {
      clearTimeout(this.timer);
    };
  }
  render() {
    const { canvasId } = this;
    return (h("canvas", { "canvas-id": canvasId, style: {
        width: '100%',
        height: '100%'
      }, onTouchStart: this.onTouchStart, onTouchMove: this.onTouchMove, onTouchEnd: this.onTouchEnd }));
  }
  get el() { return getElement(this); }
};
Canvas.style = indexCss;

export { Canvas as taro_canvas_core };
